File: desktop-mend/src/test/java/commands/UnlockTest.java
package commands;

import co.samco.mend4.core.exception.CorruptSettingsException;
import co.samco.mend4.desktop.commands.Unlock;
import co.samco.mend4.desktop.exception.SettingRequiredException;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;

import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import java.io.IOException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.interfaces.RSAPublicKey;
import java.security.spec.InvalidKeySpecException;
import java.util.Collections;

import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyString;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.*;

public class UnlockTest extends TestBase {
    private Unlock unlock;

    private ArgumentCaptor<String> errCaptor;

    private String privateKeyPath = "privateKeyPath";
    private String publicKeyPath = "publicKeyPath";

    @Before
    public void setup() {
        super.setup();
        errCaptor = ArgumentCaptor.forClass(String.class);
        when(fileResolveHelper.getPrivateKeyPath()).thenReturn(privateKeyPath);
        when(fileResolveHelper.getPublicKeyPath()).thenReturn(publicKeyPath);
        unlock = new Unlock(strings, osDao, settings, settingsHelper, log, cryptoProvider, shredHelper, fileResolveHelper, keyHelper);
    }

    @Test
    public void testCommandWithSettingsDependencies() throws IOException, SettingRequiredException {
        super.testCommandWithSettingsDependencies(unlock);
    }

    @Test
    public void wrongPassword() throws NoSuchPaddingException, InvalidKeyException, NoSuchAlgorithmException,
            IllegalBlockSizeException, BadPaddingException, InvalidKeySpecException, IOException, CorruptSettingsException {
        when(osDao.readPassword(anyString())).thenReturn(new char[0]);
        when(keyHelper.getPublicKey()).thenReturn(null);
        when(cryptoProvider.checkPassword(any(char[].class), any(String.class), any(RSAPublicKey.class)))
                .thenReturn(false);
        unlock.execute(Collections.emptyList());
        verify(err).println(eq(strings.get("Unlock.incorrectPassword")));
        verify(osDao, never()).fileExists(any());
        verify(shredHelper, never()).tryShredFile(any());
        verify(osDao, never()).writeDataToFile(any(), any());
    }

    @Test
    public void correctPasswordNoExistingKeys() throws NoSuchPaddingException, InvalidKeyException, NoSuchAlgorithmException,
            IllegalBlockSizeException, BadPaddingException, InvalidKeySpecException, IOException, CorruptSettingsException {
        correctPasswordTest();
        verify(shredHelper, never()).tryShredFile(any());
    }

    @Test
    public void correctPasswordExistingKeys() throws IOException, NoSuchAlgorithmException, InvalidKeyException,
            NoSuchPaddingException, BadPaddingException, CorruptSettingsException, InvalidKeySpecException,
            IllegalBlockSizeException {
        when(osDao.fileExists(any())).thenReturn(true);
        correctPasswordTest();
        verify(shredHelper, times(2)).tryShredFile(any());
    }

    private void correctPasswordTest() throws NoSuchPaddingException, InvalidKeyException, NoSuchAlgorithmException,
            IllegalBlockSizeException, BadPaddingException, InvalidKeySpecException,
            IOException, CorruptSettingsException {
        String password = "password";
        when(osDao.readPassword(anyString())).thenReturn(password.toCharArray());
        when(keyHelper.getPublicKey()).thenReturn(null);
        when(cryptoProvider.checkPassword(any(char[].class), any(String.class), any(RSAPublicKey.class)))
                .thenReturn(true);
        unlock.execute(Collections.emptyList());
        verify(osDao, times(2)).writeDataToFile(any(), any());
    }
}

File: desktop-mend/src/test/java/commands/UnlockTest.kt
package commands

import co.samco.mend4.core.Settings
import co.samco.mend4.core.crypto.UnlockResult
import co.samco.mend4.desktop.commands.Unlock
import org.junit.Before
import org.junit.Test
import org.mockito.ArgumentMatchers.anyString
import com.nhaarman.mockitokotlin2.KArgumentCaptor
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.eq
import com.nhaarman.mockitokotlin2.never
import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import org.apache.commons.codec.binary.Base64
import org.junit.Assert.assertEquals
import java.io.ByteArrayOutputStream
import java.io.File
import java.security.PrivateKey

class UnlockTest : TestBase() {
    private lateinit var unlock: Unlock
    private lateinit var errCaptor: KArgumentCaptor<String>
    private val privateKeyFile = File("privateKeyPath-test")
    private val publicKeyFile = File("publicKeyPath-test")

    @Before
    override fun setup() {
        super.setup()
        errCaptor = argumentCaptor()
        whenever(fileResolveHelper.privateKeyFile).thenReturn(privateKeyFile)
        whenever(fileResolveHelper.publicKeyFile).thenReturn(publicKeyFile)

        unlock = Unlock(
            strings = strings,
            settings = settings,
            settingsHelper = settingsHelper,
            log = log,
            cryptoProvider = cryptoProvider,
            shredHelper = shredHelper,
            fileResolveHelper = fileResolveHelper,
            osDao = osDao,
        )
    }

    @Test
    fun testCommandWithSettingsDependencies() {
        super.testCommandWithNoSettingsDependencies(unlock)
    }

    @Test
    fun wrongPassword() {
        whenever(osDao.readPassword(anyString())).thenReturn(CharArray(0))
        whenever(cryptoProvider.unlock(any())).thenReturn(UnlockResult.Failure)

        unlock.execute(emptyList())

        verify(err).println(eq(strings["Unlock.incorrectPassword"]))
        verify(osDao, never()).exists(any())
        verify(shredHelper, never()).tryShredFile(any())
        verify(osDao, never()).fileOutputSteam(any(), any())
    }

    @Test
    fun correctPasswordNoExistingKeys() {
        correctPasswordTest()
        verify(shredHelper, never()).tryShredFile(any())
    }

    @Test
    fun correctPasswordExistingKeys() {
        whenever(osDao.exists(any())).thenReturn(true)
        correctPasswordTest()
        verify(shredHelper, times(2)).tryShredFile(any())
    }

    private fun privateKeyForBytes(byteArray: ByteArray): PrivateKey {
        return object : PrivateKey {
            override fun getAlgorithm(): String = "algorithm"

            override fun getFormat(): String = "format"

            override fun getEncoded(): ByteArray = byteArray
        }
    }

    private fun correctPasswordTest() {
        val password = "password".toCharArray()
        whenever(osDao.readPassword(anyString())).thenReturn(password)
        whenever(cryptoProvider.unlock(eq(password)))
            .thenReturn(UnlockResult.Success(privateKeyForBytes("privateKey".toByteArray())))
        whenever(settings.getValue(eq(Settings.Name.PUBLIC_KEY)))
            .thenReturn(Base64.encodeBase64URLSafeString("publicKey".toByteArray()))

        val privateKeyBytes = ByteArrayOutputStream()
        val publicKeyBytes = ByteArrayOutputStream()
        whenever(osDao.fileOutputSteam(eq(privateKeyFile), eq(false)))
            .thenReturn(privateKeyBytes)
        whenever(osDao.fileOutputSteam(eq(publicKeyFile), eq(false)))
            .thenReturn(publicKeyBytes)

        unlock.execute(emptyList())
        verify(osDao, times(1)).fileOutputSteam(eq(privateKeyFile), eq(false))
        verify(osDao, times(1)).fileOutputSteam(eq(publicKeyFile), eq(false))
        assertEquals("privateKey", String(privateKeyBytes.toByteArray()))
        assertEquals("publicKey", String(publicKeyBytes.toByteArray()))
    }
}